library/std/src/thread/scoped.rs

```diff
@@ -35,28 +39,144 @@ pub struct Scope<'scope, 'env: 'scope> {
 #[stable(feature = "scoped_threads", since = "1.63.0")]
 pub struct ScopedJoinHandle<'scope, T>(JoinInner<'scope, T>);
 
+/// Set by the main thread to indicate that `main_thread` has a value
+/// and it is waiting to be notified by zeroing out the `sync_state`.
+const WAITING_BIT: usize = 1;
+
+/// Represents one thread spawned by the scope that is actively running.
+const ONE_RUNNING: usize = 2;
+
+/// A soft limit on the amount of concurrent spawned threads in a scope.
+///
+/// Going above this limit will panic to avoid the `sync_state` wrapping
+/// around and potetially triggering undefined behavior.
+const MAX_RUNNING: usize = usize::MAX / 2;
+
+#[derive(Default)]
 pub(super) struct ScopeData {
-    num_running_threads: AtomicUsize,
+    sync_state: AtomicUsize,
     a_thread_panicked: AtomicBool,
-    main_thread: Thread,
+    main_thread: Cell<Option<Thread>>,
+    _pinned: PhantomPinned,
 }
 
+unsafe impl Send for ScopeData {}
+unsafe impl Sync for ScopeData {}
+
 impl ScopeData {
-    pub(super) fn increment_num_running_threads(&self) {
-        // We check for 'overflow' with usize::MAX / 2, to make sure there's no
-        // chance it overflows to 0, which would result in unsoundness.
-        if self.num_running_threads.fetch_add(1, Ordering::Relaxed) > usize::MAX / 2 {
-            // This can only reasonably happen by mem::forget()'ing many many ScopedJoinHandles.
-            self.decrement_num_running_threads(false);
-            panic!("too many running threads in thread scope");
+    /// Generates an Acquire fence in relation to `sync_state`.
+    fn sync_state_acquire_fence(&self) {
+        // ThreadSanitizer does not support memory fences.
+        // To avoid false positives, issue a stub acquire load
+        // to create the happens-before edge for synchronization instead.
+        if cfg!(sanitize = "thread") {
+            self.sync_state.load(Ordering::Acquire);
+        } else {
+            crate::sync::atomic::fence(Ordering::Acquire);
         }
     }
+
+    #[inline]
+    pub(super) fn increment_num_running_threads(&self) {
+        let state = self.sync_state.fetch_add(ONE_RUNNING, Ordering::Relaxed);
+        let num_running_threads = state / ONE_RUNNING;
+        assert!(num_running_threads <= MAX_RUNNING, "too many running threads in thread scope");
+    }
+
+    #[inline]
     pub(super) fn decrement_num_running_threads(&self, panic: bool) {
         if panic {
             self.a_thread_panicked.store(true, Ordering::Relaxed);
         }
-        if self.num_running_threads.fetch_sub(1, Ordering::Release) == 1 {
-            self.main_thread.unpark();
+
+        unsafe {
+            // Move to using raw pointers instead of references as decrementing a pending count
+            // could cause the main thread to return and invalidate our ScopeData reference
+            // while it is still active and dereferenceable.
+            let self_ptr = self as *const Self;
+            drop(self);
```

Comment:
`drop(self)` here doesn't do anything, because `&Self` is `Copy`

Reply:
would the function need to take the raw pointer as the argument?

Reply:
No, as raw pointers are Copy too. I don't think there's a way to do what you want here, but it's not clearly needed?

Reply:
Should be fine for raw pointers to be Copy as they're not dereferenceable after the atomic sub/store.

Reply:
`drop(...)` does nothing to Copy types. It's not magic: https://doc.rust-lang.org/nightly/src/core/mem/mod.rs.html#974

Reply:
Oh, I see. Yes, if `decrement_num_running_threads` took `*const Self` then while `drop(...)` wouldn't do anything, you wouldn't have the issue of a dangling reference.